### PR TITLE
Add `Ecto.Model.association_loaded?`

### PR DESCRIPTION
**Pull request by paulcsmith**
A function for checking whether a passed in association is loaded
or not.

### CODE REVIEW

**Comment by josevalim** [comment]
I think it should rather be `Ecto.Associations.loaded?`. It isn't going to be used that much to justify polluting Ecto.Model, what do you think?

**Comment by paulcsmith** [comment]
I think that makes more sense.

In IRC eric suggested Ecto.Model though. What do you think @ericmj?

File: lib/ecto/model.ex [hunk 1]
```diff
@@ -130,6 +130,24 @@ defmodule Ecto.Model do
...
+
+    post = Repo.get(Post, 1)
+    Ecto.Model.association_loaded?(post.comments) # false
+    post = post |> Repo.preload(:comments)
+    Ecto.Model.association_loaded?(post.comments) # true
+
+  """
+  def association_loaded?(association) do
+    case association do
+      %{__field__: _} -> false
```
Comment (josevalim) [inline lib/ecto/model.ex]: You can match on `%Ecto.Associations.NotLoaded{}` instead!

Comment (paulcsmith) [inline lib/ecto/model.ex]: Oops! That makes sense. Thanks!

**Comment by ericmj** [comment]
I agree with @josevalim, lets go with `Ecto.Associations.loaded?`.

File: lib/ecto/associations.ex [hunk 2]
```diff
@@ -99,6 +99,24 @@ defmodule Ecto.Associations do
...
+
+    post = Repo.get(Post, 1)
+    Ecto.Associations.loaded?(post.comments) # false
+    post = post |> Repo.preload(:comments)
+    Ecto.Associations.loaded?(post.comments) # true
+
+  """
+  def loaded?(association) do
+    case association do
+      %Ecto.Associations.NotLoaded{} -> false
```
Comment (paulcsmith) [inline lib/ecto/associations.ex]: The problem is that `Ecto.Associations.NotLoaded` is not yet defined so it throws: `Ecto.Associations.NotLoaded.__struct__/0 is undefined, cannot expand struct Ecto.Associations.NotLoaded`

What's the best practice for fixing this issue?

Comment (josevalim) [inline lib/ecto/associations.ex]: You can move the struct definition up, before this module. :)

Comment (paulcsmith) [inline lib/ecto/associations.ex]: Sounds good :) I wasn't sure if I should extract each module to it's own file in an `/associations` directory to clean this up or not.

**Comment by paulcsmith** [comment]
All tests pass and suggestions have been implemented

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Pull request merged**